framework/src/modules/chain/blocks/blocks.js

```diff
@@ -0,0 +1,558 @@
+					await blocksChain.deleteLastBlock();
+					// emit event
+					this._lastBlock = await blocksChain.deleteLastBlock();
+					// emit event
+					this._isActive = false;
+					setImmediate(cb);
+					return;
+				} catch (error) {
+					this._isActive = false;
+					this.logger.error(error);
+					setImmediate(cb, error);
+					return;
+				}
+			}
+			if (blocksVerify.isForkFive(block, this._lastBlock)) {
+				this.roundsModule.fork(block, 5);
+				if (blocksVerify.isDoubleForge(block, this._lastBlock)) {
+					this.logger.warn(
+						'Delegate forging on multiple nodes',
+						block.generatorPublicKey
+					);
+				}
+				if (blocksVerify.shouldDiscardForkFive(block, this._lastBlock)) {
+					this.logger.info('Last block stands');
+					setImmediate(cb);
+					this._isActive = false;
+					return;
+				}
+				this._updateLastReceipt();
+				try {
+					const { verified, errors } = blocksVerify.normalizeAndVerify({
+						...this.constants,
+						block,
+						lastBlock: this._lastBlock,
+						exceptions: this.exceptions,
+						roundsModule: this.roundsModule,
+						slots: this.slots,
+						blockReward: this.blockReward,
+						lastNBlockIds: this._lastNBlockIds,
+					});
+					if (!verified) {
+						throw errors;
+					}
+					const deletingBlock = this._lastBlock;
+					this._lastBlock = await blocksChain.deleteLastBlock();
+					this.emit(EVENT_DELETE_BLOCK, { block: deletingBlock });
+					// emit event
+					this._lastBlock = await this._processBlock(
+						block,
+						this._lastBlock,
+						validBlock => this.broadcast(validBlock)
+					)();
+					await this._updateBroadhash();
+					this._isActive = false;
+					setImmediate(cb);
+					return;
+				} catch (error) {
+					this.logger.error(error);
+					this._isActive = false;
+					setImmediate(cb, error);
+					return;
+				}
+			}
+			// Discard received block
+			this._isActive = false;
+			setImmediate(cb);
+		});
+	}
+
+	// Process a block from syncing
+	async loadBlocksFromNetwork(blocks) {
+		this._shouldNotBeActive();
+		this._isActive = true;
+		const normalizedBlocks = blocksUtils.readDbRows(
+			blocks,
+			this.transactionManager,
+			this.genesisBlock
+		);
+		// eslint-disable-next-line no-restricted-syntax
+		for (const block of normalizedBlocks) {
+			// check if it's cleaning
+			if (this._cleaning) {
+				break;
+			}
+			// eslint-disable-next-line no-await-in-loop
+			this._lastBlock = await this._processBlock(block, this._lastBlock)();
+			// emit event
+			this._updateLastNBlocks(block);
+			this.emit(EVENT_NEW_BLOCK, { block });
+		}
+		this._isActive = false;
+		return this._lastBlock;
+	}
+
+	// Generate a block for forging
+	async generateBlock(keypair, timestamp, transactions = []) {
+		this._shouldNotBeActive();
+		this._isActive = true;
+		try {
+			const block = await blocksProcess.generateBlock({
+				keypair,
+				timestamp,
+				transactions,
+				lastBlock: this._lastBlock,
+				storage: this.storage,
+				exceptions: this.exceptions,
```

Comment:
generateBlock should never be passed with exceptions. If we need to do it due to consistency, then it should be an empty array.